jax/interpreters/partial_eval.py

```diff
@@ -1064,17 +1076,23 @@ def ensure_instantiated(inst: bool, x: Atom) -> Atom:
   out_unknowns, out_inst = unzip2(map(read, jaxpr.outvars))
   assert all(type(v) is Var for v in residuals), residuals
 
+  for x, inst, ensure_inst in zip(jaxpr.outvars, out_inst, ensure_out_inst):
+    if ensure_inst: ensure_instantiated(inst, x)
+  out_unknowns = map(op.or_, out_unknowns, ensure_out_unknowns)
+  out_inst     = map(op.or_, out_inst,     ensure_out_inst)
+
   ins_known, _ = partition_list(in_unknowns, jaxpr.invars)
   outs_known, _ = partition_list(out_unknowns, jaxpr.outvars)
   known_effects = core.join_effects(*(eqn.effects for eqn in known_eqns))
-  jaxpr_known = Jaxpr((), ins_known, [*outs_known, *residuals], known_eqns,
-                      known_effects)
+  jaxpr_known = Jaxpr(jaxpr.constvars, ins_known, [*outs_known, *residuals],
```

Comment:
Are the constvars just forwarded through into the known/staged jaxprs? I'm assuming the caller needs to DCE them away.

Reply:
Yeah, indeed I took the approach that both sides get constvars. If the caller wants to partition them between 'known' and 'staged' sides, they can be converted to invars (via `convert_constvars_jaxpr`). Similarly for DCE, constvars aren't pruned by `dce_jaxpr`, though the caller can just convert to invars. In both of these cases constvars just aren't affected.

We may want to revise that in the future, but this behavior was most consistent with the callers today.